packages/next/src/server/base-server.ts

```diff
@@ -1741,6 +1746,45 @@ export default abstract class Server<ServerOptions extends Options = Options> {
     }
   }
 
+  protected pathCouldBeIntercepted(resolvedPathname: string): boolean {
+    return (
+      isInterceptionRouteAppPath(resolvedPathname) ||
+      this.interceptionRouteRewrites?.some((rewrite) => {
+        return new RegExp(rewrite.regex).test(resolvedPathname)
+      })
+    )
+  }
+
+  protected setVaryHeader(
+    req: BaseNextRequest,
+    res: BaseNextResponse,
+    isAppPath: boolean,
+    resolvedPathname: string
+  ): void {
+    const baseVaryHeader = `${RSC_HEADER}, ${NEXT_ROUTER_STATE_TREE}, ${NEXT_ROUTER_PREFETCH_HEADER}`
+    const isRSCRequest =
+      Boolean(req.headers[RSC_HEADER.toLowerCase()]) ||
+      getRequestMeta(req, 'isRSCRequest')
+    let addedNextUrlToVary = false
+
+    if (isAppPath && this.pathCouldBeIntercepted(resolvedPathname)) {
+      // Interception route responses can vary based on the `Next-URL` header.
+      // We use the Vary header to signal this behavior to the client to properly cache the response.
+      res.setHeader('vary', `${baseVaryHeader}, ${NEXT_URL}`)
+      addedNextUrlToVary = true
+    } else if (isAppPath || isRSCRequest) {
```

Comment:
Aren't all app path's also RSC request paths 🤔?

Reply:
We discussed offline, but for posterity this is to preserve the existing behavior where we added the `Vary` header for pages RSC requests (which can happen on prefetches that occur from app router pages)